src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+                    "an HKDF ExtractThenExpand could not be initialized with "
+                    + "the given key or salt material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+            "an HKDF could not be initialized with the given KDFParameterSpec");
+    }
+
+    private static boolean isNullOrEmpty(Collection<?> c) {
+        return c == null || c.isEmpty();
+    }
+
+    private SecretKey consolidateKeyMaterial(List<SecretKey> keys)
+        throws InvalidKeyException {
+        if (keys != null && !keys.isEmpty()) {
+            ArrayList<SecretKey> localKeys = new ArrayList<>(keys);
+            if (localKeys.size() == 1) {
+                // return this element
+                SecretKey checkIt = localKeys.get(0);
+                byte[] workItemBytes = CipherCore.getKeyBytes(checkIt);
+                return new SecretKeySpec(workItemBytes, "Generic");
+            } else {
+                byte[] bb = new byte[0];
+                for (SecretKey workItem : localKeys) {
+                    byte[] workItemBytes = CipherCore.getKeyBytes(workItem);
+
+                    bb = Arrays.copyOf(bb, bb.length + workItemBytes.length);
+                    System.arraycopy(workItemBytes, 0, bb,
+                                     bb.length - workItemBytes.length,
+                                     workItemBytes.length);
+                }
+                return new SecretKeySpec(bb, "Generic");
+            }
+        } else if (keys.isEmpty()) {
+            return null;
+        } else {
+            throw new InvalidKeyException(
+                "List of keys could not be consolidated");
+        }
+    }
+
+    /**
+     * Perform the HMAC-Extract operation.
+     *
+     * @param inputKey
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract.  If no salt is to be used a
+     *     {@code null} value should be provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the
+     *     {@code HkdfParameterSpec}
+     */
+    protected byte[] hkdfExtract(SecretKey inputKey, byte[] salt)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+        hmacObj.init(new SecretKeySpec(salt, "HKDF-Salt"));
```

Comment:
Why use "HKDF-Salt" as key algorithm name? This is just internal key for the Hmac object and maybe it's safer to just use the `hmacAlgName` as the key algorithm? The key algorithm does not affect the Hmac operation anyway.

Reply:
I think Valerie is right. In case a Mac implementation is picky on the key algorithm name.

Reply:
Yes, resolved.